### DISCUSSION TITLE
How to convert `source_set.add(when)` into `if` statement? Doc is unclear

### DISCUSSION THREAD

**Discussion by hehaoqian** [mesonbuild/meson#11140] [Q&A] [answered]
I am beginner to meson build, 
and just begin my QEMU development, which uses meson

In QEMU, there are many code similar to this for conditional compilation

```
ss = import('sourceset')
ss.add(when: 'CONFIG_XXXX', if_true: files('a.c'))
# The above code works as intended
````

But how can I check if `'CONFIG_XXXX'` is true in an `if` statement,
so I can put multiple meson expression guarded by a config

What I tried:
```
if CONFIG_XXXX
endif

if 'CONFIG_XXXX'
endif

if config['CONFIG_XXXX']
endif

if configuration_data().get['CONFIG_XXXX']
endif
```

None of the above works, error during meson configure

The doc of [source_set](https://mesonbuild.com/SourceSet-module.html#source_set) says

> source_set.add([when: varnames_and_deps],
>                [if_true: sources_and_deps],
>                [if_false: list_of_alt_sources])
> varnames_and_deps is a list of conditions for the rule, which can be either strings or dependency objects 
> (a dependency object is anything that has a found() method).
> **If all the strings evaluate to true** and all dependencies are found, 
> the rule will evaluate to true; 

I am very confused that how `source_set.add(when)` resolves string into boolean value?
A raw string in meson is not directly convertible to boolean in meson,
but how is it converted to bool in `source_set.add(when)`?

I guess the string is a key to a dictionary? But how to get the dictionary?
The documentation is unclear here

The exact QEMU commit I use is https://github.com/T-head-Semi/qemu/commit/fd25f40f7d66d6fe5caffa9ef4b610434de6cf42
It uses meson 0.55.3,
and the meson commit is https://github.com/mesonbuild/meson/commit/776acd2a805c9b42b4f0375150977df42130317f

**Comment by dcbaker** [comment, marked as answer]
Those values are resolved when `source_set.apply()` is called, but are stored as strings until that time.

The basic opration is something like this:
```meson
conf = configuration_data()
conf.set('FOO_TRUE', true)
conf.set('FOO_FALSE', false)

s = source_set()
s.add(when : 'FOO_TRUE', files: files('true.c'))
s.add(when : 'FOO_FALSE', files : files('false.c'))

sources = s.apply(conf).sources()
assert(sources == files('true.c')
```

So if you want to change that into a conditional inline, you need to have whichever `configuration_data` that will be passed to apply:
```meson
conf = configuration_data()
conf.set('FOO_TRUE', true)
conf.set('FOO_FALSE', false)

if conf.get('FOO_TRUE')
  ...
endif
```
